From d8b6f5b5c6b92adff8965ab634b672e7b333f583 Mon Sep 17 00:00:00 2001
From: Henri Yandell <bayard@apache.org>
Date: Tue, 31 Oct 2006 21:34:03 +0000
Subject: [PATCH] Applying unit test and fix for #LANG-292. Also fixes a couple
 of problems with the unescape(Writer..) overload that came up

git-svn-id: https://svn.apache.org/repos/asf/jakarta/commons/proper/lang/trunk@469661 13f79535-47bb-0310-9956-ffa450edef68
---
 src/java/org/apache/commons/lang/Entities.java     | 9 ++++++++-
 src/test/org/apache/commons/lang/EntitiesTest.java | 8 ++++++++
 2 files changed, 16 insertions(+), 1 deletion(-)

diff --git a/src/java/org/apache/commons/lang/Entities.java b/src/java/org/apache/commons/lang/Entities.java
index 495c98180d4..a45e0049c28 100644
--- a/src/java/org/apache/commons/lang/Entities.java
+++ b/src/java/org/apache/commons/lang/Entities.java
@@ -847,6 +847,9 @@ public String unescape(String str) {
                             } else {
                                 entityValue = Integer.parseInt(entityName.substring(1));
                             }
+                            if (entityValue > 0xFFFF) {
+                                entityValue = -1;
+                            }
                         } catch (NumberFormatException ex) {
                             entityValue = -1;
                         }
@@ -917,13 +920,17 @@ public void unescape(Writer writer, String string) throws IOException {
                                     case 'X' :
                                     case 'x' : {
                                         entityValue = Integer.parseInt(entityContent.substring(2), 16);
+                                        break;
                                     }
                                     default : {
                                         entityValue = Integer.parseInt(entityContent.substring(1), 10);
                                     }
                                 }
+                                if (entityValue > 0xFFFF) {
+                                    entityValue = -1;
+                                }
                             } catch (NumberFormatException e) {
-                                // ignore the escaped value content
+                                entityValue = -1;
                             }
                         }
                     } else { //escaped value content is an entity name
diff --git a/src/test/org/apache/commons/lang/EntitiesTest.java b/src/test/org/apache/commons/lang/EntitiesTest.java
index fb444dde010..0e2604a1dab 100644
--- a/src/test/org/apache/commons/lang/EntitiesTest.java
+++ b/src/test/org/apache/commons/lang/EntitiesTest.java
@@ -197,5 +197,13 @@ public void testHtml40Nbsp() throws Exception
         assertEquals("&nbsp;", e.escape("\u00A0"));
     }
 
+    public void testNumberOverflow() throws Exception {
+        doTestUnescapeEntity("&#12345678;", "&#12345678;");
+        doTestUnescapeEntity("x&#12345678;y", "x&#12345678;y");
+        doTestUnescapeEntity("&#x12345678;", "&#x12345678;");
+        doTestUnescapeEntity("x&#x12345678;y", "x&#x12345678;y");
+    }
+
+
 }
 
